### DISCUSSION TITLE
Content-Type: application/x-www-form-urlencoded

### DISCUSSION THREAD

**Discussion by tuhinganguly** [Huachao/vscode-restclient#767] [Q&A]
How do I create a POST request with Content-Type: application/x-www-form-urlencoded

I have a CURL something like below:
`curl --location --request POST 'https://{{formStoreHost}}/connect/token' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--header 'Authorization: Basic {{formStoreToken}}' \
--data-urlencode 'grant_type=client_credentials' \
--data-urlencode 'scope=formstore_api'`

I get a response :
`{
  "error": "unsupported_grant_type"
}'

@Huachao can you please help look at this. I am trying to see if the tool would fit our team's requirement and I am getting stuck with these kind of requests.

**Comment by Huachao** [comment]
@tuhinganguly since `--data-urlencode` is not supported by my extension

**Reply by adamaig** [reply]
@Huachao would it be possible to add? I'm unclear from your response. Not supporting it means that when `rest-client.formParamEncodingStrategy = always` the generated code is inconsistent, so this would be expected behavior.

I'm unclear on why the `--data-urlencode` isn't implemented and want to cofirm that it is a time or capacity issue vs a different kind of constraint before looking to see if I can contribute a PR that you might accept.

[update]
I walked the code a bit and think I understand the issue. In order to generate the cURL output, the input is converted to HAR format and then run through httpsnippet, which in turn doesn't generate `--data-urlencode` outputs. 

If this is correct, then it seems like either the encoding of the request body isn't occurring when generating the cURL request, or it is being undone as it goes through the parsing and generation pipeline.

For reference, my rest-client example is configured with `"rest-client.formParamEncodingStrategy": "always"` and an example that runs into encoding issues is:

```
# the username is an email address like fname+test@gmail.com
POST https://example.com/connect/token HTTP/1.1
Content-Type: application/x-www-form-urlencoded

client_id=My.Client&client_secret=12345SoSecret&grant_type=password&username= fname+test@gmail.com&password=mypass&scope=email openid
```

The expected output would have the values of each key encoded, as illustrated here:

```
curl --request POST \
  --url https://example.com/connect/token \
  --header 'content-type: application/x-www-form-urlencoded' \
  --header 'user-agent: vscode-restclient' \
  --data client_id=MyClient \
  --data client_secret=12345SoSecret \
  --data grant_type=password \
  --data 'username=fname%2Btest%40gmail.com' \
  --data password=mypass \
  --data 'scope=openid%20email'
```

Alternatively, would you consider adding a filter instruction so that individual values could easily be encoded when desired? Maybe syntax like `{{urlencode myvar}}`?
